nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/QueryIoTDBRecord.java

```diff
@@ -0,0 +1,219 @@
+            throw new ProcessException("The IoTDB Query processor requires input connection or scheduled IoTDB query");
+        }
+        recordSetWriterFactory = context.getProperty(RECORD_WRITER_FACTORY).asControllerService(RecordSetWriterFactory.class);
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession processSession) throws ProcessException {
+
+        String query;
+        Charset charset;
+        FlowFile outgoingFlowFile;
+        int fetchSize = DEFAULT_IOTDB_FETCH_SIZE;
+        // If there are incoming connections, prepare query params from flow file
+        if (context.hasIncomingConnection()) {
+            FlowFile incomingFlowFile = processSession.get();
+
+            if (incomingFlowFile == null && context.hasNonLoopConnection()) {
+                return;
+            }
+            fetchSize = context.getProperty(IOTDB_QUERY_FETCH_SIZE).evaluateAttributeExpressions(incomingFlowFile).asInteger();
+            if ( (incomingFlowFile != null ? incomingFlowFile.getSize() : 0) == 0 ) {
+                if ( context.getProperty(IOTDB_QUERY).isSet() ) {
+                    query = context.getProperty(IOTDB_QUERY).evaluateAttributeExpressions(incomingFlowFile).getValue();
+                } else {
+                    String message = "FlowFile query is empty and no scheduled query is set";
+                    getLogger().error(message);
+                    incomingFlowFile = processSession.putAttribute(incomingFlowFile, "iotdb.error.message", message);
+                    processSession.transfer(incomingFlowFile, REL_FAILURE);
+                    return;
+                }
+            } else {
+                try {
+                    query = getQuery(processSession, incomingFlowFile);
+                } catch(IOException ioe) {
+                    final String message = String.format("Reading Query from FlowFile failed %s", incomingFlowFile);
+                    throw new ProcessException(message, ioe);
+                }
+            }
+            outgoingFlowFile = incomingFlowFile;
+        } else {
+            outgoingFlowFile = processSession.create();
+            query = context.getProperty(IOTDB_QUERY).evaluateAttributeExpressions(outgoingFlowFile).getValue();
+        }
+
+        try {
+            SessionDataSet dataSet = session.get().executeQueryStatement(query);
+            List<Map<String,Object>> result = new ArrayList<>();
+            dataSet.setFetchSize(fetchSize);
+            List<String> fieldType = dataSet.getColumnTypes();
+            List<String> fieldNames = dataSet.getColumnNames();
+            List<RecordField> recordFields = new ArrayList<>();
+            for (int i = 0; i < fieldNames.size(); i++) {
+                recordFields.add(new RecordField(fieldNames.get(i), getType(fieldType.get(i)).getDataType()));
+            }
+            final RecordSchema schema = new SimpleRecordSchema(recordFields);
```

Comment:
It looks like creating the SimpleRecordSchema could be extracted to a separate private method. Iterating over result data set could use the fields defined from the schema to maintain encapsulation.